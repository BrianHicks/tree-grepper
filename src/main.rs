use clap::Clap;
use ignore::{types, WalkBuilder, WalkState};
use std::fs;
use std::path::PathBuf;
use std::process;
use tree_sitter::{self, Query};

#[derive(Clap, Debug)]
#[clap(version = "1.0")]
struct Opts {
    /// Pattern to search for.
    pattern: String,

    /// Paths to look for files. Can be files, directories, or a mix of both.
    #[clap(default_value = ".", parse(from_os_str))]
    paths: Vec<PathBuf>,

    /// How deeply to recurse (default: no limit)
    #[clap(short('d'), long)]
    max_depth: Option<usize>,

    /// Follow symlinks
    #[clap(short('f'), long("follow"))]
    follow_links: bool,

    /// How many threads to use when loading files (default: choose automatically based on heuristics from ripgrep)
    #[clap(long, default_value = "0")]
    threads: usize,
    // TODO: add more options from https://docs.rs/ignore/0.4.16/ignore/struct.WalkBuilder.html
}

fn main() {
    let opts: Opts = Opts::parse();

    // safety checks: can we get a parser?
    if let Err(err) = parser(language_elm()) {
        eprintln!(
            "Couldn't get the parser because of an internal error: {:?}",
            err
        );
        process::exit(1);
    }

    // safety check: is the query acceptable?
    // TODO: this error type has rich enough text to make a really nice error
    // message, but this implementation ends up pretty crappy. Make it better!
    if let Err(err) = Query::new(language_elm(), &opts.pattern) {
        eprintln!("Invalid pattern: {:?}", err);
        process::exit(1);
    }

    // I *think* we should be OK to assume that there's at least one path in
    // this `opts.paths`, since there will be a default set above. This code
    // is a little incautious as a result, but a future refactor could break
    // it! Is there a better way? (e.g. making it impossible by construction
    // like `(a, Vec<a>)`?)
    let mut builder = WalkBuilder::new(opts.paths[0].clone());

    // argh! how do I iterate starting at index 1? Is this the right way?
    let mut idx = 1;
    while let Some(path) = opts.paths.get(idx) {
        builder.add(path);
        idx += 1;
    }

    // TODO: move type definitions to another function
    let mut types_builder = types::TypesBuilder::new();
    if let Err(err) = types_builder.add("elm", "*.elm") {
        eprintln!("Couldn't add Elm type: {:?}", err);
        process::exit(1);
    }
    types_builder.select("elm");

    let types = match types_builder.build() {
        Ok(t) => t,
        Err(err) => {
            eprintln!("Couldn't select file types: {:?}", err);
            process::exit(1);
        }
    };

    builder
        .follow_links(opts.follow_links)
        .max_depth(opts.max_depth)
        .threads(opts.threads)
        .types(types)
        .build_parallel()
        .run(|| {
            let mut parser = match parser(language_elm()) {
                Ok(p) => p,
                Err(_) => return Box::new(|_| WalkState::Quit),
            };

            Box::new(move |dir_entry_result| match dir_entry_result {
                Err(err) => {
                    eprintln!("Error reading path: {:}", err);
                    WalkState::Quit
                }
                Ok(dir_entry) => {
                    if dir_entry.path().is_dir() {
                        return WalkState::Continue;
                    }

                    let source = match fs::read_to_string(dir_entry.path()) {
                        Ok(s) => s,
                        Err(err) => {
                            eprintln!("Couldn't read source of {:?}: {:}", dir_entry.path(), err);
                            return WalkState::Quit;
                        }
                    };

                    let tree = match parser.parse(&source, None) {
                        Some(t) => t,
                        None => {
                            eprintln!("Couldn't parse source of {:?}", dir_entry.path());
                            return WalkState::Quit;
                        }
                    };

                    println!("{:#?}", tree);
                    WalkState::Continue
                }
            })
        });
}

// tree-sitter setup

extern "C" {
    fn tree_sitter_elm() -> tree_sitter::Language;
}

fn language_elm() -> tree_sitter::Language {
    unsafe { tree_sitter_elm() }
}

fn parser(
    language: tree_sitter::Language,
) -> Result<tree_sitter::Parser, tree_sitter::LanguageError> {
    let mut parser = tree_sitter::Parser::new();

    parser.set_language(language)?;

    Ok(parser)
}
