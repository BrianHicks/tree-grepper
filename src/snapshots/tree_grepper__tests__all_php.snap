---
source: src/main.rs
assertion_line: 373
expression: "call(&[\"tree-grepper\", \"-q\", \"php\", \"(_)\", \"--format=pretty-json\", \"--sort\",\n       \"--no-gitignore\", \"vendor/tree-sitter-php/test/highlight\"])"

---
[
  {
    "file": "vendor/tree-sitter-php/test/highlight/keywords.php",
    "file_type": "php",
    "matches": [
      {
        "kind": "program",
        "name": "query",
        "text": "<?php\n// <- tag\n\nif ($a) {}\n// <- keyword\n\nwhile ($b) {}\n// <- keyword\n\nWHILE ($b) {}\n// <- keyword\n\ndo { } while ($c);\n// <- keyword\n//     ^ keyword\n\ntry {} catch (Exception $e) {}\n// <- keyword\n//     ^ keyword\n\nfunction a() {}\n// <- keyword\n\nclass A {}\n// <- keyword\n\nthrow new Exception(\"oh\");\n// <- keyword\n//    ^ keyword\n\nfunction b(\n  int $a,\n  // <- type.builtin\n\n  string $b,\n  // <- type.builtin\n\n  Person $e\n  // ^ type\n): Dog {}\n// ^ type\n\ninterface T {}\n// ^ keyword\n\ntrait T {}\n// ^ keyword\n",
        "start": {
          "row": 1,
          "column": 1
        },
        "end": {
          "row": 48,
          "column": 1
        }
      },
      {
        "kind": "php_tag",
        "name": "query",
        "text": "<?php",
        "start": {
          "row": 1,
          "column": 1
        },
        "end": {
          "row": 1,
          "column": 6
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- tag",
        "start": {
          "row": 2,
          "column": 1
        },
        "end": {
          "row": 2,
          "column": 10
        }
      },
      {
        "kind": "if_statement",
        "name": "query",
        "text": "if ($a) {}",
        "start": {
          "row": 4,
          "column": 1
        },
        "end": {
          "row": 4,
          "column": 11
        }
      },
      {
        "kind": "parenthesized_expression",
        "name": "query",
        "text": "($a)",
        "start": {
          "row": 4,
          "column": 4
        },
        "end": {
          "row": 4,
          "column": 8
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$a",
        "start": {
          "row": 4,
          "column": 5
        },
        "end": {
          "row": 4,
          "column": 7
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 4,
          "column": 6
        },
        "end": {
          "row": 4,
          "column": 7
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 4,
          "column": 9
        },
        "end": {
          "row": 4,
          "column": 11
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 5,
          "column": 1
        },
        "end": {
          "row": 5,
          "column": 14
        }
      },
      {
        "kind": "while_statement",
        "name": "query",
        "text": "while ($b) {}",
        "start": {
          "row": 7,
          "column": 1
        },
        "end": {
          "row": 7,
          "column": 14
        }
      },
      {
        "kind": "parenthesized_expression",
        "name": "query",
        "text": "($b)",
        "start": {
          "row": 7,
          "column": 7
        },
        "end": {
          "row": 7,
          "column": 11
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$b",
        "start": {
          "row": 7,
          "column": 8
        },
        "end": {
          "row": 7,
          "column": 10
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "b",
        "start": {
          "row": 7,
          "column": 9
        },
        "end": {
          "row": 7,
          "column": 10
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 7,
          "column": 12
        },
        "end": {
          "row": 7,
          "column": 14
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 8,
          "column": 1
        },
        "end": {
          "row": 8,
          "column": 14
        }
      },
      {
        "kind": "while_statement",
        "name": "query",
        "text": "WHILE ($b) {}",
        "start": {
          "row": 10,
          "column": 1
        },
        "end": {
          "row": 10,
          "column": 14
        }
      },
      {
        "kind": "parenthesized_expression",
        "name": "query",
        "text": "($b)",
        "start": {
          "row": 10,
          "column": 7
        },
        "end": {
          "row": 10,
          "column": 11
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$b",
        "start": {
          "row": 10,
          "column": 8
        },
        "end": {
          "row": 10,
          "column": 10
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "b",
        "start": {
          "row": 10,
          "column": 9
        },
        "end": {
          "row": 10,
          "column": 10
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 10,
          "column": 12
        },
        "end": {
          "row": 10,
          "column": 14
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 11,
          "column": 1
        },
        "end": {
          "row": 11,
          "column": 14
        }
      },
      {
        "kind": "do_statement",
        "name": "query",
        "text": "do { } while ($c);",
        "start": {
          "row": 13,
          "column": 1
        },
        "end": {
          "row": 13,
          "column": 19
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{ }",
        "start": {
          "row": 13,
          "column": 4
        },
        "end": {
          "row": 13,
          "column": 7
        }
      },
      {
        "kind": "parenthesized_expression",
        "name": "query",
        "text": "($c)",
        "start": {
          "row": 13,
          "column": 14
        },
        "end": {
          "row": 13,
          "column": 18
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$c",
        "start": {
          "row": 13,
          "column": 15
        },
        "end": {
          "row": 13,
          "column": 17
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "c",
        "start": {
          "row": 13,
          "column": 16
        },
        "end": {
          "row": 13,
          "column": 17
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 14,
          "column": 1
        },
        "end": {
          "row": 14,
          "column": 14
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//     ^ keyword",
        "start": {
          "row": 15,
          "column": 1
        },
        "end": {
          "row": 15,
          "column": 17
        }
      },
      {
        "kind": "try_statement",
        "name": "query",
        "text": "try {} catch (Exception $e) {}",
        "start": {
          "row": 17,
          "column": 1
        },
        "end": {
          "row": 17,
          "column": 31
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 17,
          "column": 5
        },
        "end": {
          "row": 17,
          "column": 7
        }
      },
      {
        "kind": "catch_clause",
        "name": "query",
        "text": "catch (Exception $e) {}",
        "start": {
          "row": 17,
          "column": 8
        },
        "end": {
          "row": 17,
          "column": 31
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "Exception",
        "start": {
          "row": 17,
          "column": 15
        },
        "end": {
          "row": 17,
          "column": 24
        }
      },
      {
        "kind": "named_type",
        "name": "query",
        "text": "Exception",
        "start": {
          "row": 17,
          "column": 15
        },
        "end": {
          "row": 17,
          "column": 24
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "Exception",
        "start": {
          "row": 17,
          "column": 15
        },
        "end": {
          "row": 17,
          "column": 24
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$e",
        "start": {
          "row": 17,
          "column": 25
        },
        "end": {
          "row": 17,
          "column": 27
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "e",
        "start": {
          "row": 17,
          "column": 26
        },
        "end": {
          "row": 17,
          "column": 27
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 17,
          "column": 29
        },
        "end": {
          "row": 17,
          "column": 31
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 18,
          "column": 1
        },
        "end": {
          "row": 18,
          "column": 14
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//     ^ keyword",
        "start": {
          "row": 19,
          "column": 1
        },
        "end": {
          "row": 19,
          "column": 17
        }
      },
      {
        "kind": "function_definition",
        "name": "query",
        "text": "function a() {}",
        "start": {
          "row": 21,
          "column": 1
        },
        "end": {
          "row": 21,
          "column": 16
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 21,
          "column": 10
        },
        "end": {
          "row": 21,
          "column": 11
        }
      },
      {
        "kind": "formal_parameters",
        "name": "query",
        "text": "()",
        "start": {
          "row": 21,
          "column": 11
        },
        "end": {
          "row": 21,
          "column": 13
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 21,
          "column": 14
        },
        "end": {
          "row": 21,
          "column": 16
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 22,
          "column": 1
        },
        "end": {
          "row": 22,
          "column": 14
        }
      },
      {
        "kind": "class_declaration",
        "name": "query",
        "text": "class A {}",
        "start": {
          "row": 24,
          "column": 1
        },
        "end": {
          "row": 24,
          "column": 11
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "A",
        "start": {
          "row": 24,
          "column": 7
        },
        "end": {
          "row": 24,
          "column": 8
        }
      },
      {
        "kind": "declaration_list",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 24,
          "column": 9
        },
        "end": {
          "row": 24,
          "column": 11
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 25,
          "column": 1
        },
        "end": {
          "row": 25,
          "column": 14
        }
      },
      {
        "kind": "expression_statement",
        "name": "query",
        "text": "throw new Exception(\"oh\");",
        "start": {
          "row": 27,
          "column": 1
        },
        "end": {
          "row": 27,
          "column": 27
        }
      },
      {
        "kind": "throw_expression",
        "name": "query",
        "text": "throw new Exception(\"oh\")",
        "start": {
          "row": 27,
          "column": 1
        },
        "end": {
          "row": 27,
          "column": 26
        }
      },
      {
        "kind": "object_creation_expression",
        "name": "query",
        "text": "new Exception(\"oh\")",
        "start": {
          "row": 27,
          "column": 7
        },
        "end": {
          "row": 27,
          "column": 26
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "Exception",
        "start": {
          "row": 27,
          "column": 11
        },
        "end": {
          "row": 27,
          "column": 20
        }
      },
      {
        "kind": "arguments",
        "name": "query",
        "text": "(\"oh\")",
        "start": {
          "row": 27,
          "column": 20
        },
        "end": {
          "row": 27,
          "column": 26
        }
      },
      {
        "kind": "argument",
        "name": "query",
        "text": "\"oh\"",
        "start": {
          "row": 27,
          "column": 21
        },
        "end": {
          "row": 27,
          "column": 25
        }
      },
      {
        "kind": "encapsed_string",
        "name": "query",
        "text": "\"oh\"",
        "start": {
          "row": 27,
          "column": 21
        },
        "end": {
          "row": 27,
          "column": 25
        }
      },
      {
        "kind": "string",
        "name": "query",
        "text": "oh",
        "start": {
          "row": 27,
          "column": 22
        },
        "end": {
          "row": 27,
          "column": 24
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- keyword",
        "start": {
          "row": 28,
          "column": 1
        },
        "end": {
          "row": 28,
          "column": 14
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//    ^ keyword",
        "start": {
          "row": 29,
          "column": 1
        },
        "end": {
          "row": 29,
          "column": 16
        }
      },
      {
        "kind": "function_definition",
        "name": "query",
        "text": "function b(\n  int $a,\n  // <- type.builtin\n\n  string $b,\n  // <- type.builtin\n\n  Person $e\n  // ^ type\n): Dog {}",
        "start": {
          "row": 31,
          "column": 1
        },
        "end": {
          "row": 40,
          "column": 10
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "b",
        "start": {
          "row": 31,
          "column": 10
        },
        "end": {
          "row": 31,
          "column": 11
        }
      },
      {
        "kind": "formal_parameters",
        "name": "query",
        "text": "(\n  int $a,\n  // <- type.builtin\n\n  string $b,\n  // <- type.builtin\n\n  Person $e\n  // ^ type\n)",
        "start": {
          "row": 31,
          "column": 11
        },
        "end": {
          "row": 40,
          "column": 2
        }
      },
      {
        "kind": "simple_parameter",
        "name": "query",
        "text": "int $a",
        "start": {
          "row": 32,
          "column": 3
        },
        "end": {
          "row": 32,
          "column": 9
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "int",
        "start": {
          "row": 32,
          "column": 3
        },
        "end": {
          "row": 32,
          "column": 6
        }
      },
      {
        "kind": "primitive_type",
        "name": "query",
        "text": "int",
        "start": {
          "row": 32,
          "column": 3
        },
        "end": {
          "row": 32,
          "column": 6
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$a",
        "start": {
          "row": 32,
          "column": 7
        },
        "end": {
          "row": 32,
          "column": 9
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 32,
          "column": 8
        },
        "end": {
          "row": 32,
          "column": 9
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- type.builtin",
        "start": {
          "row": 33,
          "column": 3
        },
        "end": {
          "row": 33,
          "column": 21
        }
      },
      {
        "kind": "simple_parameter",
        "name": "query",
        "text": "string $b",
        "start": {
          "row": 35,
          "column": 3
        },
        "end": {
          "row": 35,
          "column": 12
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "string",
        "start": {
          "row": 35,
          "column": 3
        },
        "end": {
          "row": 35,
          "column": 9
        }
      },
      {
        "kind": "primitive_type",
        "name": "query",
        "text": "string",
        "start": {
          "row": 35,
          "column": 3
        },
        "end": {
          "row": 35,
          "column": 9
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$b",
        "start": {
          "row": 35,
          "column": 10
        },
        "end": {
          "row": 35,
          "column": 12
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "b",
        "start": {
          "row": 35,
          "column": 11
        },
        "end": {
          "row": 35,
          "column": 12
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- type.builtin",
        "start": {
          "row": 36,
          "column": 3
        },
        "end": {
          "row": 36,
          "column": 21
        }
      },
      {
        "kind": "simple_parameter",
        "name": "query",
        "text": "Person $e",
        "start": {
          "row": 38,
          "column": 3
        },
        "end": {
          "row": 38,
          "column": 12
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "Person",
        "start": {
          "row": 38,
          "column": 3
        },
        "end": {
          "row": 38,
          "column": 9
        }
      },
      {
        "kind": "named_type",
        "name": "query",
        "text": "Person",
        "start": {
          "row": 38,
          "column": 3
        },
        "end": {
          "row": 38,
          "column": 9
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "Person",
        "start": {
          "row": 38,
          "column": 3
        },
        "end": {
          "row": 38,
          "column": 9
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$e",
        "start": {
          "row": 38,
          "column": 10
        },
        "end": {
          "row": 38,
          "column": 12
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "e",
        "start": {
          "row": 38,
          "column": 11
        },
        "end": {
          "row": 38,
          "column": 12
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// ^ type",
        "start": {
          "row": 39,
          "column": 3
        },
        "end": {
          "row": 39,
          "column": 12
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "Dog",
        "start": {
          "row": 40,
          "column": 4
        },
        "end": {
          "row": 40,
          "column": 7
        }
      },
      {
        "kind": "named_type",
        "name": "query",
        "text": "Dog",
        "start": {
          "row": 40,
          "column": 4
        },
        "end": {
          "row": 40,
          "column": 7
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "Dog",
        "start": {
          "row": 40,
          "column": 4
        },
        "end": {
          "row": 40,
          "column": 7
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 40,
          "column": 8
        },
        "end": {
          "row": 40,
          "column": 10
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// ^ type",
        "start": {
          "row": 41,
          "column": 1
        },
        "end": {
          "row": 41,
          "column": 10
        }
      },
      {
        "kind": "interface_declaration",
        "name": "query",
        "text": "interface T {}",
        "start": {
          "row": 43,
          "column": 1
        },
        "end": {
          "row": 43,
          "column": 15
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "T",
        "start": {
          "row": 43,
          "column": 11
        },
        "end": {
          "row": 43,
          "column": 12
        }
      },
      {
        "kind": "declaration_list",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 43,
          "column": 13
        },
        "end": {
          "row": 43,
          "column": 15
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// ^ keyword",
        "start": {
          "row": 44,
          "column": 1
        },
        "end": {
          "row": 44,
          "column": 13
        }
      },
      {
        "kind": "trait_declaration",
        "name": "query",
        "text": "trait T {}",
        "start": {
          "row": 46,
          "column": 1
        },
        "end": {
          "row": 46,
          "column": 11
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "T",
        "start": {
          "row": 46,
          "column": 7
        },
        "end": {
          "row": 46,
          "column": 8
        }
      },
      {
        "kind": "declaration_list",
        "name": "query",
        "text": "{}",
        "start": {
          "row": 46,
          "column": 9
        },
        "end": {
          "row": 46,
          "column": 11
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// ^ keyword",
        "start": {
          "row": 47,
          "column": 1
        },
        "end": {
          "row": 47,
          "column": 13
        }
      }
    ]
  },
  {
    "file": "vendor/tree-sitter-php/test/highlight/literals.php",
    "file_type": "php",
    "matches": [
      {
        "kind": "program",
        "name": "query",
        "text": "<?php\n\n\necho <<<OMG\n  something\nOMG\n// <- string\n\necho true, TRUE, false, FALSE\n//    ^ constant.builtin\n//          ^ constant.builtin\n//               ^ constant.builtin\n//                      ^ constant.builtin\n\necho PI_314\n//   ^ constant\n",
        "start": {
          "row": 1,
          "column": 1
        },
        "end": {
          "row": 17,
          "column": 1
        }
      },
      {
        "kind": "php_tag",
        "name": "query",
        "text": "<?php",
        "start": {
          "row": 1,
          "column": 1
        },
        "end": {
          "row": 1,
          "column": 6
        }
      },
      {
        "kind": "echo_statement",
        "name": "query",
        "text": "echo <<<OMG\n  something\nOMG\n// <- string",
        "start": {
          "row": 4,
          "column": 1
        },
        "end": {
          "row": 7,
          "column": 13
        }
      },
      {
        "kind": "heredoc",
        "name": "query",
        "text": " <<<OMG\n  something\nOMG",
        "start": {
          "row": 4,
          "column": 5
        },
        "end": {
          "row": 6,
          "column": 4
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "// <- string",
        "start": {
          "row": 7,
          "column": 1
        },
        "end": {
          "row": 7,
          "column": 13
        }
      },
      {
        "kind": "ERROR",
        "name": "query",
        "text": "echo true, TRUE, false, FALSE\n//    ^ constant.builtin\n//          ^ constant.builtin\n//               ^ constant.builtin\n//                      ^ constant.builtin\n\necho PI_314",
        "start": {
          "row": 9,
          "column": 1
        },
        "end": {
          "row": 15,
          "column": 12
        }
      },
      {
        "kind": "boolean",
        "name": "query",
        "text": "true",
        "start": {
          "row": 9,
          "column": 6
        },
        "end": {
          "row": 9,
          "column": 10
        }
      },
      {
        "kind": "sequence_expression",
        "name": "query",
        "text": "TRUE, false, FALSE",
        "start": {
          "row": 9,
          "column": 12
        },
        "end": {
          "row": 9,
          "column": 30
        }
      },
      {
        "kind": "boolean",
        "name": "query",
        "text": "TRUE",
        "start": {
          "row": 9,
          "column": 12
        },
        "end": {
          "row": 9,
          "column": 16
        }
      },
      {
        "kind": "sequence_expression",
        "name": "query",
        "text": "false, FALSE",
        "start": {
          "row": 9,
          "column": 18
        },
        "end": {
          "row": 9,
          "column": 30
        }
      },
      {
        "kind": "boolean",
        "name": "query",
        "text": "false",
        "start": {
          "row": 9,
          "column": 18
        },
        "end": {
          "row": 9,
          "column": 23
        }
      },
      {
        "kind": "boolean",
        "name": "query",
        "text": "FALSE",
        "start": {
          "row": 9,
          "column": 25
        },
        "end": {
          "row": 9,
          "column": 30
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//    ^ constant.builtin",
        "start": {
          "row": 10,
          "column": 1
        },
        "end": {
          "row": 10,
          "column": 25
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//          ^ constant.builtin",
        "start": {
          "row": 11,
          "column": 1
        },
        "end": {
          "row": 11,
          "column": 31
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//               ^ constant.builtin",
        "start": {
          "row": 12,
          "column": 1
        },
        "end": {
          "row": 12,
          "column": 36
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//                      ^ constant.builtin",
        "start": {
          "row": 13,
          "column": 1
        },
        "end": {
          "row": 13,
          "column": 43
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "PI_314",
        "start": {
          "row": 15,
          "column": 6
        },
        "end": {
          "row": 15,
          "column": 12
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//   ^ constant",
        "start": {
          "row": 16,
          "column": 1
        },
        "end": {
          "row": 16,
          "column": 16
        }
      }
    ]
  },
  {
    "file": "vendor/tree-sitter-php/test/highlight/types.php",
    "file_type": "php",
    "matches": [
      {
        "kind": "program",
        "name": "query",
        "text": "<?php\n\n\nfunction a(array $b, Something $c) {\n  //       ^ type.builtin\n  //                  ^ type\n  echo (int) $foo;\n  //     ^ type.builtin\n}\n",
        "start": {
          "row": 1,
          "column": 1
        },
        "end": {
          "row": 10,
          "column": 1
        }
      },
      {
        "kind": "php_tag",
        "name": "query",
        "text": "<?php",
        "start": {
          "row": 1,
          "column": 1
        },
        "end": {
          "row": 1,
          "column": 6
        }
      },
      {
        "kind": "function_definition",
        "name": "query",
        "text": "function a(array $b, Something $c) {\n  //       ^ type.builtin\n  //                  ^ type\n  echo (int) $foo;\n  //     ^ type.builtin\n}",
        "start": {
          "row": 4,
          "column": 1
        },
        "end": {
          "row": 9,
          "column": 2
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 4,
          "column": 10
        },
        "end": {
          "row": 4,
          "column": 11
        }
      },
      {
        "kind": "formal_parameters",
        "name": "query",
        "text": "(array $b, Something $c)",
        "start": {
          "row": 4,
          "column": 11
        },
        "end": {
          "row": 4,
          "column": 35
        }
      },
      {
        "kind": "simple_parameter",
        "name": "query",
        "text": "array $b",
        "start": {
          "row": 4,
          "column": 12
        },
        "end": {
          "row": 4,
          "column": 20
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "array",
        "start": {
          "row": 4,
          "column": 12
        },
        "end": {
          "row": 4,
          "column": 17
        }
      },
      {
        "kind": "primitive_type",
        "name": "query",
        "text": "array",
        "start": {
          "row": 4,
          "column": 12
        },
        "end": {
          "row": 4,
          "column": 17
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$b",
        "start": {
          "row": 4,
          "column": 18
        },
        "end": {
          "row": 4,
          "column": 20
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "b",
        "start": {
          "row": 4,
          "column": 19
        },
        "end": {
          "row": 4,
          "column": 20
        }
      },
      {
        "kind": "simple_parameter",
        "name": "query",
        "text": "Something $c",
        "start": {
          "row": 4,
          "column": 22
        },
        "end": {
          "row": 4,
          "column": 34
        }
      },
      {
        "kind": "type_list",
        "name": "query",
        "text": "Something",
        "start": {
          "row": 4,
          "column": 22
        },
        "end": {
          "row": 4,
          "column": 31
        }
      },
      {
        "kind": "named_type",
        "name": "query",
        "text": "Something",
        "start": {
          "row": 4,
          "column": 22
        },
        "end": {
          "row": 4,
          "column": 31
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "Something",
        "start": {
          "row": 4,
          "column": 22
        },
        "end": {
          "row": 4,
          "column": 31
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$c",
        "start": {
          "row": 4,
          "column": 32
        },
        "end": {
          "row": 4,
          "column": 34
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "c",
        "start": {
          "row": 4,
          "column": 33
        },
        "end": {
          "row": 4,
          "column": 34
        }
      },
      {
        "kind": "compound_statement",
        "name": "query",
        "text": "{\n  //       ^ type.builtin\n  //                  ^ type\n  echo (int) $foo;\n  //     ^ type.builtin\n}",
        "start": {
          "row": 4,
          "column": 36
        },
        "end": {
          "row": 9,
          "column": 2
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//       ^ type.builtin",
        "start": {
          "row": 5,
          "column": 3
        },
        "end": {
          "row": 5,
          "column": 26
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//                  ^ type",
        "start": {
          "row": 6,
          "column": 3
        },
        "end": {
          "row": 6,
          "column": 29
        }
      },
      {
        "kind": "echo_statement",
        "name": "query",
        "text": "echo (int) $foo;",
        "start": {
          "row": 7,
          "column": 3
        },
        "end": {
          "row": 7,
          "column": 19
        }
      },
      {
        "kind": "cast_expression",
        "name": "query",
        "text": "(int) $foo",
        "start": {
          "row": 7,
          "column": 8
        },
        "end": {
          "row": 7,
          "column": 18
        }
      },
      {
        "kind": "cast_type",
        "name": "query",
        "text": "int",
        "start": {
          "row": 7,
          "column": 9
        },
        "end": {
          "row": 7,
          "column": 12
        }
      },
      {
        "kind": "variable_name",
        "name": "query",
        "text": "$foo",
        "start": {
          "row": 7,
          "column": 14
        },
        "end": {
          "row": 7,
          "column": 18
        }
      },
      {
        "kind": "name",
        "name": "query",
        "text": "foo",
        "start": {
          "row": 7,
          "column": 15
        },
        "end": {
          "row": 7,
          "column": 18
        }
      },
      {
        "kind": "comment",
        "name": "query",
        "text": "//     ^ type.builtin",
        "start": {
          "row": 8,
          "column": 3
        },
        "end": {
          "row": 8,
          "column": 24
        }
      }
    ]
  }
]
